ui/components/app/qr-hardware-popover/enhanced-reader.js

```diff
@@ -0,0 +1,67 @@
+import React, { useEffect, useMemo, useState } from 'react';
+import { BarcodeFormat, DecodeHintType } from '@zxing/library';
+import { BrowserQRCodeReader } from '@zxing/browser';
+import log from 'loglevel';
+import PropTypes from 'prop-types';
+import { MILLISECOND } from '../../../../shared/constants/time';
+import Spinner from '../../ui/spinner';
+
+const EnhancedReader = ({ handleScan }) => {
+  const [canplay, setCanplay] = useState(false);
+  const codeReader = useMemo(() => {
+    const hint = new Map();
+    hint.set(DecodeHintType.POSSIBLE_FORMATS, [BarcodeFormat.QR_CODE]);
+    return new BrowserQRCodeReader(hint, {
+      delayBetweenScanAttempts: MILLISECOND * 100,
+      delayBetweenScanSuccess: MILLISECOND * 100,
+    });
+  }, []);
+
+  useEffect(() => {
+    const videoElem = document.getElementById('video');
+    const canplayListener = () => {
+      setCanplay(true);
+    };
+    videoElem.addEventListener('canplay', canplayListener);
+    const promise = codeReader.decodeFromVideoDevice(
+      undefined,
+      'video',
+      (result) => {
+        if (result) {
+          handleScan(result.getText());
+        }
+      },
+    );
+    return () => {
+      videoElem.removeEventListener('canplay', canplayListener);
+      promise
+        .then((controls) => {
```

Comment:
is this the correct way to do async `useEffect`

Reply:
this is okay, it could use async/await, but this will work

Reply:
@soralit Could you use async/await here, with a try-catch?

Reply:
Hi @kumavis @danjm , thanks to your review. Not sure do you means passing an async callback to useEffect, like `useEffect(async () => {}, [])`, but in my knowledge of React, it is not recommended because React expects an effect hook to return undefined or a synchronous clean-up function, using async callback will make the clean-up work fail when the component unmount. Ignore me if I misunderstand you.

Reply:
On the other hand, if you means to create a async function and call it immediately in the effect hook like this:
```
const [controls, setControls] = useState(null);
useEffect(() => {
    if(controls === null) {
      const setupDecoder = async () => {
        try {
          const controls = await codeReader.decodeFromVideoDevice(
            undefined,
            'video',
            (result) => {
              if (result) {
                handleScan(result.getText());
              }
            },
          )
          setControls(controls)
        } catch (e) {
          log.info(e)
        }
      }
      setupDecoder();
    }
    return () => {
      if(controls) {
        controls.stop();
      }
    };
  }, [controls]);
```
I just tried this style of code and find it will make the QR code reader not working correctly in this situation: 
> 1. Open the popover and try read QR code
> 2. Close the popover before camera was loaded
> 3. The camera now is unable to close. We will see the indicator of camera is always open until we close the tab. 
Actually we need the state `controls` to stop the camera when component is unmount, but in the step 2 the state `controls` was not settled yet when component was unmount, but the async function `decodeFromVideoDevice` is still running even the component is unmounted.